geode-core/src/distributedTest/java/org/apache/geode/cache/query/dunit/PdxMultiThreadQueryDUnitTest.java

```diff
@@ -176,45 +177,78 @@ public void testClientServerQueryUsingRemoteQueryService()
         new int[] {port0, port1, port2}, true);
 
-    try {
-      client.invoke(() -> {
-        TestPdxSerializable.throwExceptionOnDeserialization = true;
+    client.invoke(() -> {
+      try {
+        TestObjectThrowsPdxSerializationException.throwExceptionOnDeserialization = true;
         QueryService remoteQueryService = (PoolManager.find(poolName)).getQueryService();
         logger.info("### Executing Query on server: " + queryString[1]);
         Query query = remoteQueryService.newQuery(queryString[1]);
-        SelectResults<TestPdxSerializable> selectResults = (SelectResults) query.execute();
-        assertEquals(numberOfEntries, selectResults.size());
-      });
-    } finally {
-      client.invoke(() -> {
-        assertEquals(false, TestPdxSerializable.throwExceptionOnDeserialization);
-        // the failed try will increment numInstance once
-        assertEquals(numberOfEntries + 1, TestPdxSerializable.numInstance.get());
-        TestPdxSerializable.numInstance.set(0);
-      });
-    }
+        SelectResults<TestObjectThrowsPdxSerializationException> selectResults =
+            (SelectResults) query.execute();
+        assertThat(selectResults.size()).isEqualTo(numberOfEntries);
+        // the 2 failed try incremented numInstance
+        assertThat(numberOfEntries + 2)
+            .isEqualTo(TestObjectThrowsPdxSerializationException.numInstance.get());
+      } finally {
+        assertThat(TestObjectThrowsPdxSerializationException.throwExceptionOnDeserialization)
+            .isFalse();
+        TestObjectThrowsPdxSerializationException.numInstance.set(0);
+      }
+    });
+  }
+
+  @Test
+  public void testRetryFailedWithServerConnectivityException() throws CacheException {
+    // create pdx instance at servers
+    server0.invoke(() -> {
+      Region<Object, Object> region = getRootRegion().getSubregion(regionName);
+      for (int i = 0; i < numberOfEntries; i++) {
+        region.put("key-" + i, new TestObjectThrowsPdxSerializationException());
+      }
+    });
+
+    // Create client pool with only 2 servers to test that retry will not run forever
+    createPool(client, poolName, new String[] {hostName, hostName},
+        new int[] {port0, port1}, true);
+
+    client.invoke(() -> {
+      try {
+        TestObjectThrowsPdxSerializationException.throwExceptionOnDeserialization = true;
+        QueryService remoteQueryService = (PoolManager.find(poolName)).getQueryService();
+        logger.info("### Executing Query on server: " + queryString[1]);
+        Query query = remoteQueryService.newQuery(queryString[1]);
+        SelectResults<TestObjectThrowsPdxSerializationException> selectResults =
+            (SelectResults) query.execute();
+        fail("Expect ServerConnectivityException");
+      } catch (ServerConnectivityException sce) {
+        logger.info("Expect ServerConnectivityException after tried 2 servers");
+      } finally {
```

Comment:
Rather than using `fail()` and catching the expected exception here, it would be better to use:
```
        assertThatThrownBy(query::execute).isInstanceOf(ServerConnectivityException.class);
```
Doing this would mean that you don't need to use the try/finally here too, which would simplify the test slightly.

Reply:
done